### PR TITLE
MDEV-16372 ER_BASE64_DECODE_ERROR upon replaying binary log with syst…

### PR DESCRIPTION
**Pull request by midenok**
…em table (#503)

mysqlbinlog verbose mode in case of multi-event BINLOG statement caches events for verbose output and prints it after BINLOG statement is finished.

BINLOG statement must not be splitted into multiple statements because row events require Table_map_event.

Tested on all suites.

[Fixes tempesta-tech/mariadb#500]

This code is submitted under the BSD-new license.

### CODE REVIEW

**Comment by midenok** [comment]
Versioned UPDATE puts 2 different row events (UPDATE_ROWS_EVENT_V1, WRITE_ROWS_EVENT_V1) into single statement. It's not reproducible just for multi-row statement consisting from same event like:
```sql
insert into t1 values (1), (2);
```
Result
```
+BINLOG '
+kpswWxMBAAAALQAAAJMCAAAAACEAAAAAAAEABHRlc3QAAnQxAAEDAAHRS5yD
+kpswWxcBAAAAKwAAAL4CAAAAACEAAAAAAAEAAf/+AQAAAP4CAAAAem7ftQ==
+'/*!*/;
+### INSERT INTO `test`.`t1`  
+### SET
+###   @1=1
+### INSERT INTO `test`.`t1`  
+### SET
+###   @1=2
+# Number of rows: 2
+# at 702
+#180625 10:36:50 server id 1  end_log_pos 775 CRC32 0xb1be0e92         Query   thread_id=5     exec_time=0     error_code=0
```

**Comment by vuvova** [comment]
1. why is it not reproducible for multi-row events consisting of events of the same type?
2. you can have mix events with, say, `insert t1 values (1), (2) on duplicate key update a=3`

**Comment by midenok** [comment]
1. Not reproducible with multi-row INSERT, because it is in fact a single event with multiple rows. See this loop:
```c++
#0  Rows_log_event::print_verbose (this=0xa40c50, file=0x7fffffffbe20, print_event_info=0x7fffffffb7c8) at /home/midenok/src/mariadb/upstream/src/sql/log_event.cc:3617
#1  0x00000000004204fd in Log_event::print_base64 (this=0xa409a0, file=0x7fffffffbe20, print_event_info=0x7fffffffb7c8, more=false) at /home/midenok/src/mariadb/upstream/src/sql/log_event.cc:3867
#2  0x000000000042c3ee in Rows_log_event::print_helper (this=0xa409a0, file=0x7ffff67f4720 <_IO_2_1_stdout_>, print_event_info=0x7fffffffb7c8, name=0x4e37c5 "Write_rows") at /home/midenok/src/mariadb/upstream/src/sql/log_event.cc:11902
#3  0x000000000042d734 in Write_rows_log_event::print (this=0xa409a0, file=0x7ffff67f4720 <_IO_2_1_stdout_>, print_event_info=0x7fffffffb7c8) at /home/midenok/src/mariadb/upstream/src/sql/log_event.cc:13407
#4  0x000000000040af5a in print_base64 (print_event_info=0x7fffffffb7c8, ev=0xa409a0) at /home/midenok/src/mariadb/upstream/src/client/mysqlbinlog.cc:892
#5  0x000000000040b15c in print_row_event (print_event_info=0x7fffffffb7c8, ev=0xa409a0, table_id=24, is_stmt_end=true) at /home/midenok/src/mariadb/upstream/src/client/mysqlbinlog.cc:976
#6  0x0000000000409348 in process_event (print_event_info=0x7fffffffb7c8, ev=0xa409a0, pos=556, logname=0x7fffffffc7c8 "/tmp/upstream.000004") at /home/midenok/src/mariadb/upstream/src/client/mysqlbinlog.cc:1467
#7  0x00000000004364ae in dump_local_log_entries (print_event_info=0x7fffffffb7c8, logname=0x7fffffffc7c8 "/tmp/upstream.000004") at /home/midenok/src/mariadb/upstream/src/client/mysqlbinlog.cc:2975
#8  0x000000000040cc92 in dump_log_entries (logname=0x7fffffffc7c8 "/tmp/upstream.000004") at /home/midenok/src/mariadb/upstream/src/client/mysqlbinlog.cc:2202
#9  0x000000000040c4e0 in main (argc=0, argv=0xa346e8) at /home/midenok/src/mariadb/upstream/src/client/mysqlbinlog.cc:3135
```
```c++
3614      for (const uchar *value= m_rows_buf; value < m_rows_end; )
3615      {
3616        size_t length;
3617        print_event_info->row_events++;
3618
3619        if (my_b_printf(file, "### %s %`s.%`s\n",
3620                        sql_command,
3621                        map->get_db_name(), map->get_table_name()))
....
3631        /* Print the first image */
3632        if (!(length= print_verbose_one_row(file, td, print_event_info,
3633                                      &m_cols, value,
3634                                      (const uchar*) sql_clause1)))
3635          goto err;
3636        value+= length;
```

2. Right. Replaced test case with non-versioned one.

File: mysql-test/main/mysqlbinlog_row_minimal.result [hunk 1]
```diff
@@ -330,3 +330,17 @@ ROLLBACK /* added by mysqlbinlog */;
...
+FLUSH BINARY LOGS;
+CREATE TABLE t1 (pk INT PRIMARY KEY);
+INSERT INTO t1 VALUES (1);
+INSERT INTO t1 VALUES (1), (2) ON DUPLICATE KEY UPDATE pk= pk + 10;
+FLUSH BINARY LOGS;
+DROP TABLE t1;
+SELECT * FROM t1;
+pk
+2
+11
```
Comment (an3l) [inline mysql-test/main/mysqlbinlog_row_minimal.result]: Should this result file be updated ?

Comment (midenok) [reply mysql-test/main/mysqlbinlog_row_minimal.result]: Can you be more specific please?

Comment (an3l) [reply mysql-test/main/mysqlbinlog_row_minimal.result]: Wrong observation, sorry :) .

**Comment by vuvova** [comment]
please, rebase it on top of the latest 10.3.

**Comment by midenok** [comment]
Conflict with 5d48ea7d07b4. Probably @andrelkin can help to solve this better.

**Comment by andrelkin** [comment]
Jira MDEV-16372 is closed so does this PR.

**Pull request closed without merging**